Fix the update of link and suggestions in LinkControl

ralucaStan opened the pull request:
When switching between two anchors in the same text block, the text input from the edit link window will not show the correct value.
See video with bug here: https://www.loom.com/share/d18dba4f4c0c43b0adb33f943e77ce0f

<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
When switching between two links, the LinkControl component does not unmount.
The value prop will change but the component does not react to this change and does
not show the updated link & suggestions.

This PR does 2 things:
- it accounts for the value change inside of LinkControl and updates it's internal state
- it modifies the update of the suggestions when the value changes. The component only handled a new empty value, showing the initial suggestions if the user was to remove the value inside the input. But because the value can change and not be empty, we need to update the suggestions for this new non-empty value.


## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
- go to a page and add two links in the same text block
- go to the first link and click it to see the edit link window
- click the Edit button and see the text input with the URL and the suggestions list
- without clicking anything else, click the second link
- observe the values in the text input and the suggestion

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ralucaStan committed:
Fix the update of link and suggestions in LinkControl

When switching between two links, the LinkControl component does not unmount.
The value prop will change but the component does not react to this change and does
not show the updated link & suggestions.

getdave commented:
Ill take a look

getdave commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -84,12 +84,13 @@ class URLInput extends Component {
 		if (
 			prevProps.value !== value &&
-			this.shouldShowInitialSuggestions()
+			! this.props.disableSuggestions &&
+			! this.isUpdatingSuggestions &&
+			! this.props.__experimentalShowInitialSuggestions
```
Could you explain this change and why it's necessary? This change removes documentation in terms of:

1. The code comment explaining what's happening.
2. A well named function call which self-documents the check logic.

ralucaStan replied:
The intention here was to bypass `shouldShowInitialSuggestions`, which was returning false because of `__experimentalShowInitialSuggestions`, and just copy its logic regarding the display of suggestions. 
I see now that there was a misunderstanding from my side regarding what initial suggestions meant so the `! this.props.__experimentalShowInitialSuggestions` part of my change does not make sense.

I am not sure how to proceed further here as I don't understand why `__experimentalShowInitialSuggestions` is set to false inside shouldShowInitialSuggestions
https://github.com/WordPress/gutenberg/blob/bd514b86b0c7866f4c92f0a395e57f660f0f7d22/packages/block-editor/src/components/url-input/index.js#L116

getdave replied:
> I am not sure how to proceed further here as I don't understand why __experimentalShowInitialSuggestions is set to false inside shouldShowInitialSuggestions

As far as I can see it isn't necessarily being set to `false` here. Rather `false` is used the _default_ value if `__experimentalShowInitialSuggestions` is _not_ provided as a prop (ie: is `undefined`). See [MDN for more details](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#default_values).

I still don't quite understand why you need to bypass the `this.shouldShowInitialSuggestions()` logic here? What was it about that which was breaking the fix you've tried to introduce here?

ralucaStan replied:
Hi @getdave,
I am sorry for the late reply.

> What was it about that which was breaking the fix you've tried to introduce here?

You can see in the video below the behavior if I don't bypass `this.shouldShowInitialSuggestions()`:
when I click on the second link I see the suggestions from the first link. 
https://www.loom.com/share/79dedfab3b6942458d2b6fa6d2c9e34b

Let's look  at what `this.shouldShowInitialSuggestions()` returns: 
```
return 
(
! this.isUpdatingSuggestions && // suggestions aren't loaded
__experimentalShowInitialSuggestions && 
! ( value && value.length ) && // there is no value in the input
! ( suggestions && suggestions.length ) // there are no suggestions loaded
)
```
Please note that this would return true only when the user selects a text and clicks Add link.

In our case, when clicking the second link, `this.shouldShowInitialSuggestions()` always returns false because:
- there is a value in the input, namely the 2nd link
- there are suggestions available, from the 1st link.

What happens is that when I open the 1st link, the suggestions are updated [with the onFocus callback](https://github.com/WordPress/gutenberg/blob/bd514b86b0c7866f4c92f0a395e57f660f0f7d22/packages/block-editor/src/components/url-input/index.js#L226-L241). When clicking on the 2nd link, onFocus is no longer called and  the suggestions aren't updated in [componentDidUpdate](https://github.com/WordPress/gutenberg/blob/bd514b86b0c7866f4c92f0a395e57f660f0f7d22/packages/block-editor/src/components/url-input/index.js#L88-L93) because of `this.shouldShowInitialSuggestions`

getdave commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -171,6 +171,16 @@ function LinkControl( {
 		isEndingEditWithFocus.current = false;
 	}, [ isEditingLink ] );
 
+	/**
+	 * Update the state value internalInputValue if the url value changes
+	 * for example when clicking on another anchor
+	 */
```
👍 👍 👍 👍  👏 👏 👏 👏 👏

getdave commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -171,6 +171,16 @@ function LinkControl( {
 
+	/**
+	 * Update the state value internalInputValue if the url value changes
+	 * for example when clicking on another anchor
```
```suggestion
	 * for example when clicking on another element which renders a `<LinkControl>`
```

getdave replied:
Might be worth documenting that this "sync state with props" is necessary because the form UI is designed not to "commit" it's changes to the controlled `value` prop unless you hit "submit". Therefore it needs to maintain an internal state value for the input which therefore must be synced with the prop.

getdave commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -84,12 +84,13 @@ class URLInput extends Component {
+			! this.props.__experimentalShowInitialSuggestions
 		) {
-			this.updateSuggestions();
+			this.updateSuggestions( value.trim() );
```
Could you remind me as to the value of `trim()` here?

ralucaStan replied:
All the other calls for `this.updateSuggestions` are made with `value.trim()`.
Ideally, the trimming should be done inside `updateSuggestions`, so that the function could just be called with value: `this.updateSuggestions(value)`.

getdave replied:
Ok let's leave "as is" in this PR and address moving the `.trim()` operation to the `updateSuggestions` method in a follow up PR. Would you be able to handle that or would you like me to take care of it?

getdave replied:
Ideally, the trimming should be done inside updateSuggestions, so that the function could just be called with value: this.updateSuggestions(value).

As of this PR https://github.com/WordPress/gutenberg/pull/35060/files#diff-6a61cdc612f2906b7754c38eb8d671d60cf4c4a07ecb4a283792a3d450947b7cR143 that is now what happens. 

So we can drop this change...

ralucaStan replied:
Thank you for the update. I've removed the trim from the argument. 
Let me know if you wanna sync regarding the next steps for this PR.

getdave reviewed (changes requested):
**Update**: just had a thought here. I think the perceived "delay" in the update is due to the `URLInput` re-fetching the search suggestions. 

https://user-images.githubusercontent.com/444434/129208437-b7dbc22d-1a1a-47ce-8fe6-ee5e54a39a19.mov

It's actually not doing a network request because the `value.url` is identified as being "URL-like" so it ends up in a direct entry handler

https://github.com/WordPress/gutenberg/blob/29d286fa6e9e2e62c429ff9f5f4025f561e70a7d/packages/block-editor/src/components/link-control/use-search-handler.js#L22-L47

This handler resolves with a `Promise`. That's why we see the delay in the update.

I guess we could:

1. Find a way to reset the suggestions before we do the "fetching" logic.
2. Deliberating introduce a delay in the `handleDirectEntry` promise resolution to force the UI to go into a loading state.

---------

Thanks for catching this bug and for the PR. I encounter the same problem working on https://github.com/WordPress/gutenberg/pull/33849 and it's really important we get it solved.

I tested it and for the most part it seemed to work well. However, it is possible to observe the component updating after the fact once you switch to the 2nd link.

In the video below you can clearly observe the link in the input changing from X -> Y as I switch between them. Ideally this wouldn't happen.


https://user-images.githubusercontent.com/444434/129205780-31d6198d-d2ee-4394-90bd-087ac6269222.mp4



I think perhaps we should consider force remounting the component based on the value of the URL.

We could try to do this by adding a `key` prop to the component and assigning it's value to `value.url`. That way if the value changes the component will force remount.

I'd like to get a second opinion on this - perhaps from @kevin940726?

The reason this is _not_ happening at the moment is - I believe - due to the fact that `Popover` doesn't remount. As a result any child component doesn't re-mount either. Thus we get stale states.

You'll know this is true because if you fully unmount `<LinkControl>` (for example by clicking away from the text entirely and then clicking back on the next link) then all the state will be updated correctly.

kevin940726 commented:
Yeah, I feel like using `key` to force a remount might be the best way forward if there's no performance issue. We can wrap `<LinkControl>` with a wrapper component so that it's not required for users to update them. Something like this:

```jsx
function LinkControlWrapper(props) {
  return <LinkControl {...props} key={props.value} />
}
```

Not sure about the delay thing you mentioned though 😅.

getdave reviewed (changes requested):
I've thought about this some more and whilst this change doens't fully resolve the issue (as described in my previous review) it does move us towards a better state.

We can look to tackle the other lagging visual update of the search _result_ in a follow up PR.

If you could address the feedback here then we can look to merge this.

getdave commented:
> Yeah, I feel like using `key` to force a remount might be the best way forward if there's no performance issue. 

Let's tackle this in a follow up PR. I believe this PR moves us towards a better state.

getdave commented:
I haven't forgotten about this one. Just trying out some other options.

ralucaStan committed:
Remove trim from function argument for updateSuggestions.

Trimming gets done in the function itself, since PR #35060.

getdave commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -84,12 +84,13 @@ class URLInput extends Component {
-			this.shouldShowInitialSuggestions()
+			! this.props.disableSuggestions &&
+			! this.isUpdatingSuggestions &&
+			! this.props.__experimentalShowInitialSuggestions
```
```suggestion
			! this.props.disableSuggestions &&
			! this.isUpdatingSuggestions
```

In my testing I don't think we need `! this.props.__experimentalShowInitialSuggestions`. 

Why?

Because this property is designed to determine whether not to show suggestions immediately (before any typing) upon mounting the control. If you turn the `__experimentalShowInitialSuggestions` to `true` then you will see that when you create a new link suggestions immediately appear as soon as the Link UI opens.

You can try this by adding the property `showInitialSuggestions={ true }` to:

https://github.com/WordPress/gutenberg/blob/576f977c405d0d5d5ead0a66c6a8d9c1bef16a06/packages/format-library/src/link/inline.js#L224

All the `<LinkControl>` tests continue to pass so I'm 🤞 that this won't have knock on effects.

getdave reviewed (approved):
I've left one change that I think should be made but other than that I'm happy to approve this.

You'll need another careful rebase.

Thanks for your patience here. I've taken far too long to get to this.

ralucaStan committed:
Update packages/block-editor/src/components/url-input/index.js

Co-authored-by: Dave Smith <getdavemail@gmail.com>

getdave commented:
@tellthemachines We'd also appreciate it if you could review this one 🙇

ralucaStan committed:
Take into consideration the case when the value changes with an empty string

This was removed by the first commits, but we need to take into consideration that we also want to show the initial suggestions when the new value is empty

ralucaStan commented:
> 

> I've left one change that I think should be made but other than that I'm happy to approve this.
> 
> You'll need another careful rebase.
> 
> Thanks for your patience here. I've taken far too long to get to this.

@getdave I have pushed the changes that I thing fix this issue and updated the description. 
With my last change I handled the update of the suggestions when the value changes in the URLInput component. The component only handled a new empty value, showing the initial suggestions, if  if the user was to remove the value inside the input (only when __experimentalShowInitialSuggestions=true). But because the value can change and not be empty, we need to update the suggestions for this new non-empty value case.

The branch is also rebased on the latest trunk

@tellthemachines this is ready for review. Ping me with anything

ralucaStan commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -84,12 +87,19 @@ class URLInput extends Component {
 			}, 100 );
 		}
 
-		// Only attempt an update on suggestions if the input value has actually changed.
```
The component only handled a new empty value, showing the initial suggestions if the user was to remove the value inside the input. But because the value can change and not be empty, we need to update the suggestions for this new non-empty value.

ralucaStan commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -187,14 +187,22 @@ function LinkControl( {
+		 */
+		if ( value?.url ) {
+			setInternalInputValue( value.url );
+		}
```
Account for the value change inside of LinkControl and updates it's internal state

tellthemachines reviewed (approved):
This is working well ✅ 

Ideally the popover shouldn't open in edit mode at all, but I think #34742 addresses that, so let's get this one in to start with!

ralucaStan committed:
Merge branch 'WordPress:trunk' into fix/link-control-value-update

getdave commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -84,12 +87,19 @@ class URLInput extends Component {
 			! this.isUpdatingSuggestions
 		) {
-			this.updateSuggestions( value );
+			if ( value && value.length ) {
```
```suggestion
			if ( value?.length ) {
```

getdave reviewed (approved):
Great work. LGTM :shipit:

ralucaStan committed:
Update packages/block-editor/src/components/url-input/index.js

Co-authored-by: Dave Smith <getdavemail@gmail.com>

noisysocks commented:
This doesn't need the back port label as it landed in 11.9.
